File: services/inventory_service/models/keg_asset.py
"""
KegAsset model - Barriles como activos fijos individuales.
"""
from sqlalchemy import Column, String, Integer, Boolean, DateTime, ForeignKey, Index
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.postgresql import UUID
from datetime import datetime
import uuid
from database import Base


class KegAsset(Base):
    """
    Barril como activo fijo individual con FSM.
    
    Cada barril es único y se trackea durante todo su ciclo de vida.
    """
    __tablename__ = "keg_assets"
    
    # Primary key (UUID)
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    
    # Identificación física
    serial_number = Column(String(50), nullable=False, unique=True, index=True)
    rfid_tag = Column(String(100), nullable=True, unique=True, index=True)
    
    # Especificaciones
    size_liters = Column(Integer, nullable=False)  # 20, 30, 50, 60
    material = Column(String(50), nullable=False, default="STAINLESS_STEEL")
    keg_type = Column(String(50), nullable=False, default="SANKE_D")
    
    # Estado actual (FSM)
    current_state = Column(String(20), nullable=False, default="EMPTY", index=True)
    production_batch_id = Column(Integer, nullable=True, index=True)
    
    # Propiedad
    ownership = Column(String(20), nullable=False, default="OWN")
    guest_brewery_id = Column(Integer, nullable=True)
    
    # Ciclo de vida
    cycle_count = Column(Integer, nullable=False, default=0)
    last_cleaned_at = Column(DateTime, nullable=True)
    last_filled_at = Column(DateTime, nullable=True)
    
    # Ubicación y cliente
    current_location = Column(String(100), nullable=True)
    client_id = Column(Integer, nullable=True, index=True)
    
    # Status flags
    is_active = Column(Boolean, nullable=False, default=True)
    needs_maintenance = Column(Boolean, nullable=False, default=False)
    
    # Audit
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    transitions = relationship("KegTransition", back_populates="keg", order_by="desc(KegTransition.timestamp)")
    
    # Indexes
    __table_args__ = (
        Index('idx_keg_state_location', 'current_state', 'current_location'),
        Index('idx_keg_batch', 'production_batch_id'),
    )
    
    def __repr__(self):
        return (
            f"<KegAsset("
            f"serial='{self.serial_number}', "
            f"state={self.current_state}, "
            f"size={self.size_liters}L"
            f")>"
        )
    
    @property
    def is_available(self) -> bool:
        """Check if keg is available for use."""
        return self.is_active and self.current_state in ["EMPTY", "CLEAN"]
    
    @property
    def is_filled(self) -> bool:
        """Check if keg has beer."""
        return self.current_state in ["FULL", "TAPPED", "IN_CLIENT", "IN_TRANSIT"]

File: desert_brew_app/lib/features/inventory/domain/entities/keg_asset.dart
import 'package:equatable/equatable.dart';

/// Keg FSM states.
/// EMPTY → CLEANING → CLEAN → FULL → DELIVERED → IN_CLIENT → RETURNED → CLEANING
enum KegState {
  empty,
  cleaning,
  clean,
  full,
  delivered,
  inClient,
  returned,
  maintenance;

  factory KegState.fromString(String s) {
    switch (s.toUpperCase()) {
      case 'EMPTY':
        return KegState.empty;
      case 'CLEANING':
        return KegState.cleaning;
      case 'CLEAN':
        return KegState.clean;
      case 'FULL':
        return KegState.full;
      case 'DELIVERED':
        return KegState.delivered;
      case 'IN_CLIENT':
        return KegState.inClient;
      case 'RETURNED':
        return KegState.returned;
      case 'MAINTENANCE':
        return KegState.maintenance;
      default:
        return KegState.empty;
    }
  }

  String get displayName {
    switch (this) {
      case KegState.empty:
        return 'Vacío';
      case KegState.cleaning:
        return 'Limpiando';
      case KegState.clean:
        return 'Limpio';
      case KegState.full:
        return 'Lleno';
      case KegState.delivered:
        return 'Entregado';
      case KegState.inClient:
        return 'En Cliente';
      case KegState.returned:
        return 'Devuelto';
      case KegState.maintenance:
        return 'Mantenimiento';
    }
  }

  bool get isAvailable =>
      this == KegState.empty || this == KegState.clean;
  bool get isFilled =>
      this == KegState.full || this == KegState.delivered || this == KegState.inClient;
}

/// Keg domain entity (UUID-keyed physical asset).
class KegAsset extends Equatable {
  const KegAsset({
    required this.id,
    required this.serialNumber,
    required this.sizeLiters,
    required this.kegType,
    required this.currentState,
    required this.ownership,
    required this.cycleCount,
    required this.isActive,
    required this.needsMaintenance,
    required this.createdAt,
    required this.updatedAt,
    this.rfidTag,
    this.productionBatchId,
    this.clientId,
    this.currentLocation,
    this.lastCleanedAt,
    this.lastFilledAt,
    this.isAvailable = false,
    this.isFilled = false,
    this.guestBreweryId,
  });

  final String id; // UUID as string
  final String serialNumber;
  final int sizeLiters;
  final String kegType;   // SANKE_D, EURO_KOMB, CORNY_PIN, CORNY_BALL
  final KegState currentState;
  final String ownership; // OWN | GUEST
  final int cycleCount;
  final bool isActive;
  final bool needsMaintenance;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String? rfidTag;
  final int? productionBatchId;
  final int? clientId;
  final String? currentLocation;
  final DateTime? lastCleanedAt;
  final DateTime? lastFilledAt;
  final bool isAvailable;
  final bool isFilled;
  final int? guestBreweryId;

  @override
  List<Object?> get props => [id, serialNumber, currentState];
}
